src/celeritas/em/model/GammaNuclearModel.cc

```diff
@@ -98,5 +112,64 @@ void GammaNuclearModel::step(CoreParams const&, CoreStateDevice&) const
 }
 #endif
 
+//---------------------------------------------------------------------------//
+/*!
+ * Build CHIPS gamma-nuclear element cross sections using G4GammaNuclearXS
+ * above the upper energy limit of G4PARTICLEXS/gamma (IAEA) data. The cross
+ * sections are derived from the parameterization developed by M. V. Kossov
+ * (CERN/ITEP Moscow) in the high energy region (106 MeV < E < 50 GeV) and
+ * from a Reggeon-based parameterization in the ultra-high-energy region
+ * (E > 50 GeV). G4GammaNuclearXS uses CHIPS (G4PhotoNuclearCrossSection)
+ * above 150 MeV and performs linear interpolation between the upper energy
+ * limit of G4PARTICLEXS/gamma (IAEA) data and 150 MeV.
+ */
+GammaNuclearModel::result_type
+GammaNuclearModel::calc_chips_xs(AtomicNumber atomic_number,
+                                 double emin,
+                                 double emax) const
+{
+    CELER_EXPECT(atomic_number);
+
+    result_type result;
+
+    // Tabulate cross sections using separate parameterizations for the high
+    // energy region (emin < E < 50 GeV) and the ultra high energy region up
+    // to the maximum valid energy (emax). The numbers of bins are chosen to
+    // adequately capture both the parameterized points (224 bins from 106 MeV
+    // to 50 GeV) and the calculations used in G4PhotoNuclearCrossSection,
+    // which can be made configurable if needed (TODO). Note that the linear
+    // interpolation between the upper limit of the IAEA cross-section data
+    // and 150 MeV, as used in G4GammaNuclearXS, is also included in this
+    // tabulation.
+    double const emid = helper_->max_high_energy();
+    size_type nbin_high = 250;
+    size_type nbin_ultra = 50;
+
+    result.x.resize(nbin_high + nbin_ultra);
+    result.y.resize(nbin_high + nbin_ultra);
+
+    // Bin sizes in logarithmic energy scale
+    double const de_high = (std::log(emid) - std::log(emin)) / (nbin_high - 1);
+    double const de_ultra = (std::log(emax) - std::log(emid))
+                            / (nbin_ultra - 1);
+
+    // Tabulate the cross section from emin to emax
+    MmSqMicroXs xs;
+    for (size_type i = 0; i < nbin_high + nbin_ultra; ++i)
+    {
+        double energy
+            = (i < nbin_high)
+                  ? std::exp(std::log(emin) + de_high * i)
+                  : std::exp(std::log(emid) + de_ultra * (i - nbin_high + 1));
+
+        result.x[i] = energy;
+        xs.value()
+            = helper_->GammaNuclearElementXS(energy, atomic_number.get());
```

Comment:
```suggestion
        MmSqMicroXs xs{helper_->GammaNuclearElementXS(energy, atomic_number.get())};
```
but if you're going to make a wrapper to Geant4's helper, you should ensure it has units.

Reply:
Updated.